fs/fs.go

```diff
@@ -0,0 +1,644 @@
+					WritesMerged:    fs.DiskStats.WritesMerged,
+					SectorsWritten:  fs.DiskStats.SectorsWritten,
+					WriteTime:       fs.DiskStats.WriteTime,
+					IoInProgress:    fs.DiskStats.IoInProgress,
+					IoTime:          fs.DiskStats.IoTime,
+					WeightedIoTime:  fs.DiskStats.WeightedIoTime,
+				}
+				if devName := getDeviceName(fs.Device); len(devName) > 0 {
+					fsStats[devName] = stats
+				} else {
+					fmt.Fprintf(os.Stderr, "Unknown device name")
+					fsStats["unknown"] = stats
+				}
+			}
+		}
+	} else {
+		fmt.Fprintf(os.Stderr, "Os.Stat failed: %v; no fs stats will be available for container %v", err, container.ID)
+	}
+
+	return fsStats, nil
+}
+
+func NewFsInfo(storageDriver string) (FsInfo, error) {
+	mounts, err := mount.GetMounts()
+	if err != nil {
+		return nil, err
+	}
+	fsInfo := &RealFsInfo{
+		partitions: make(map[string]partition, 0),
+		labels:     make(map[string]string, 0),
+		dmsetup:    &defaultDmsetupClient{},
+	}
+
+	fsInfo.addSystemRootLabel(mounts)
+
+	supportedFsType := map[string]bool{
+		// all ext systems are checked through prefix.
+		"btrfs": true,
+		"xfs":   true,
+		"zfs":   true,
+	}
+	for _, mount := range mounts {
+		var Fstype string
+		if !strings.HasPrefix(mount.Fstype, "ext") && !supportedFsType[mount.Fstype] {
+			continue
+		}
+		// Avoid bind mounts.
+		if _, ok := fsInfo.partitions[mount.Source]; ok {
+			continue
+		}
+		if mount.Fstype == "zfs" {
+			Fstype = mount.Fstype
+		}
+		fsInfo.partitions[mount.Source] = partition{
+			fsType:     Fstype,
+			mountpoint: mount.Mountpoint,
+			major:      uint(mount.Major),
+			minor:      uint(mount.Minor),
+		}
+	}
+
+	return fsInfo, nil
+}
+
+// addSystemRootLabel attempts to determine which device contains the mount for /.
+func (self *RealFsInfo) addSystemRootLabel(mounts []*mount.Info) {
+	for _, m := range mounts {
+		if m.Mountpoint == "/" {
+			self.partitions[m.Source] = partition{
+				fsType:     m.Fstype,
+				mountpoint: m.Mountpoint,
+				major:      uint(m.Major),
+				minor:      uint(m.Minor),
+			}
+			self.labels[LabelSystemRoot] = m.Source
+			return
+		}
+	}
+}
+
+// updateContainerImagesPath compares the mountpoints with possible container image mount points; if a match is found,
+// the label is added to the partition.
+func (self *RealFsInfo) updateContainerImagesPath(label string, mounts []*mount.Info, containerImagePaths map[string]struct{}) {
+	var useMount *mount.Info
+	for _, m := range mounts {
+		if _, ok := containerImagePaths[m.Mountpoint]; ok {
+			if useMount == nil || (len(useMount.Mountpoint) < len(m.Mountpoint)) {
+				useMount = m
+			}
+		}
+	}
+	if useMount != nil {
+		self.partitions[useMount.Source] = partition{
+			fsType:     useMount.Fstype,
+			mountpoint: useMount.Mountpoint,
+			major:      uint(useMount.Major),
+			minor:      uint(useMount.Minor),
+		}
+		self.labels[label] = useMount.Source
+	}
+}
+
+func (self *RealFsInfo) GetDeviceForLabel(label string) (string, error) {
```

Comment:
exported method ``RealFsInfo.GetDeviceForLabel`` should have comment or be unexported